### DISCUSSION TITLE
Icons for dark mode

### DISCUSSION THREAD

**Discussion by thahnen** [eclipse-platform/eclipse.platform.ui#2514] [Q&A] [answered]
Hey everyone,

I'm trying to find any information on whether there is a way to implement icons for dark mode (the standard one, not a plug-in) alongside the ones for the light mode in plug-ins.

If this is supported, is it supported from the `plugin.xml` files or only programmatically? If not, is there a plan to change that at some point? We currently have more or less light mode focussed icons for our plug-in but we are switching and they ain't looking good in dark mode.

Best,
Tobias

**Comment by vogella** [comment, marked as answer]
This is unfortunately not yet implemented. An attempt has been started here https://github.com/eclipse-platform/eclipse.platform.ui/pull/14

Maybe other work is ongoing, @BeckerWdf are you aware of work in this area?

**Reply by thahnen** [reply]
Hi @vogella,

thanks for the response. I'd be happy for something like this to be picked up again!

Best,
Tobias

**Reply by HeikoKlare** [reply]
We had a discussion about this topic at the community day in July 2023 There is definitely some demand for this in some of the contributing companies and we concluded with initial ideas on how to properly address the topic. Still, no one had time to pick the topic up so far, but I know that it's on the list of priorities for some of us.

**Reply by laeubi** [reply]
I have already investigated at this topic a while back you can find some things around this here:
- https://github.com/eclipse-platform/eclipse.platform.images/issues/81
- https://github.com/eclipse-pde/eclipse.pde/discussions/1315

basically it is possible but needs to be implemented. If this is crucial to your business and likes to speed up the development you can [contact me](https://xn--lubisoft-0za.gmbh/en/#opensource) to enter a contract about this specific issue or if you need dedicated enterprise support.

Of course a PR to fix the problem would also be welcome as it seems you have developers at hand that are familiar with Eclipse RCP, see the [contribution guides here](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md) and if Eclipse is an important Technology for you you maybe even want to join the [Eclipse IDE Workingroup](https://www.eclipse.org/membership/#tab-membership) that also regularly funds development in critical areas or [consider a funding](https://www.eclipse.org/sponsor/ide/) to improve Eclipse IDE in general.

**Reply by thahnen** [reply]
Hey @laeubi,

The issue is not mission-critical, but thanks for the offer to lend a helping hand :+1: We will change the icons to work on both dark and light modes, and that will be it.

---

*The next part is a bit off-topic but related to the contribution part you mentioned.*

Currently, the only people familiar with Eclipse RCP and so on are a colleague who was moved to another team and myself. So from that and a discussion with some people in the company, there is no drive to contribute to the Eclipse IDE, which I personally find sad, but I can understand it due to the lack of developers with that skillset :disappointed:

I speak I on the other hand am mostly (like 90%) of my day and night working with Eclipse and would like to contribute even more. Still, I find it a bit hard due to the extensive ecosystem of Eclipse projects and even Eclipse IDE-related projects where to help out the best without knowing precisely the internals and so on.
Also, getting into the community, in general, feels like a big step into the unknown because the community mostly consists of very dedicated and passionate long-termers, even though this might not be because of ill intent to appear this way! I also don't want to offend someone by saying this but I hope it is understandable :bowing_man: 

I'm mostly familiar (which sounds odd "only" using Eclipse and developing plug-ins and applications for 4 years now compared to people working on it for 20+ years - crazy) with Eclipse Tycho, RedDeer, PDE, and RCP and creating custom IDEs from scratch.
So the million dollar question from my side would be: Where to start apart from that and all the legal stuff ^^

Best,
Tobias

**Reply by vogella** [reply]
@thahnen if you are looking for a super easy way to contribute, you could help updating / modernizing the FAQ https://github.com/eclipse-platform/eclipse.platform/blob/master/docs/FAQ/The_Official_Eclipse_FAQs.md

Other than that small code enhancements should also be a nice way to starting. Maybe fix one compiler warning or Sonar warning, send a pull request and see how this goes?

**Reply by thahnen** [reply]
Hey @vogella,
that is a great idea. I will have a look at the FAQ, has been quite some time since I've seen these pages ^^
And true, the Sonar issues can be annoying (but I'm *somehow legally* obligated to say they're useful :smile:) and most of them actually need to be fixed! That is also a good starting point, I will write that down!

**Reply by vogella** [reply]
@thahnen start as small as possible to learn the process, it can be quite frustrating if you send lots of hours only to learn that someone does  not like a particular fix.

**Reply by mickaelistria** [reply]
@thahnen I believe that for newcomers, the most motivating approach is to try to fix a small issue you face often, or to implement a small improvement in a workflow you use daily. Once you have identified such, it can be reported to the tracker and you can start trying to contribute. The project has some good resources (as pointed above) to get started, but usually if you know PDE you're just ready enough to develop Eclipse Platform: you import the bundle you'd like to change in your workspace and work on it as usual, that's all (not even need for particular target platform if you have the default one).
Some typical example can be improving some button position or labels, or a sentence you find more complex than necessary, making sure some interesting elements are reachable through Ctrl+3...
The good thing with such change is that you'll be able to see it part of the next build just after it's merged, and you'll realize million of other people will benefit from it. From here, you'll most likely realize the workflows are not so hard and will be addicted to the feeling of helping million other develoeprs, so you'll contribute more and more.

**Reply by vogella** [reply]
@thahnen easiest fix ever: remove https://github.com/eclipse-platform/eclipse.platform/blob/master/docs/FAQ/FAQ_What_is_an_Eclipse_Innovation_Grant.md and link from https://github.com/eclipse-platform/eclipse.platform/blob/master/docs/FAQ/The_Official_Eclipse_FAQs.md#The-Eclipse-Community

**Reply by thahnen** [reply]
@vogella that was really easy, if you like, you can also review the [PR](https://github.com/eclipse-platform/eclipse.platform/pull/1617)

**Reply by HannesWell** [reply]
> I speak I on the other hand am mostly (like 90%) of my day and night working with Eclipse and would like to contribute even more. Still

From what you have said, I think you have the ideal profile for an Eclipse contributor. Having even multiple-years of experience in building Eclipse based applications and plugins is perfect and you just have to learn the process. But in your PDE contribution you already managed that well and we got your nice improvement. Everything, especially the details can be learned on demand. And you are more than welcome to continue with that. :)

We (developer) community-mentors, i.e. @opcoach and myself, are working on a landing page for new Eclipse IDE contributors that should explain all aspects and should guide those who interested to contribute to the 'Eclipse IDE' towards that goal:
- https://github.com/eclipse-ide/

We will also plan to unify the existing information in all four Eclipse SDK organizations with that documentation.
If you are interested, please have a look. Maybe it's already a bit too late for you but it would still be interesting to get your feedback if that would have been helpful or what could be improved. We are also in the process of setting up a feedback-ub at (but that's really brand new and not yet fully configured):
- https://github.com/eclipse-ide/contributing

We'll also consider the feedback from this discussion.

> So the million dollar question from my side would be: Where to start apart from that and all the legal stuff ^^

That's a good question, but I think the previous answers are already good. I just want to second that fixing sonar issues is often a good starting point. In general many things can be cleaned up in the Eclipse code base and the Sonar-Lint Plugin can really help to easily identify those places. Personally I'm a big fan of your plugin, it has already prevented me from implementing bugs in changes, helps to make the code cleaner (IMO) and I now and then even learned new things from following the advises. Indeed the proposed solutions are sometimes not suitable, but in most cases I think it makes the code better.

**Reply by thahnen** [reply]
Hi @HannesWell, thanks as Well for your Kind words and the positive Feedback on SonarLint! 🙏

On the other notes about summarizing and in some Form standardizing the documentation for newbies across the Stack Seems super Helpful to me and I will Most certainly have a Look!
This is One Point I noticed as Well: Most projects have different Approaches which is understandable but makes it complex. Lifting a Bit of the complexity and having some degree of Common ground is more than promising to Jump the fence I guess!

**Comment by laeubi** [comment]
@thahnen at the very minimum, you could setup and SDK like described here:

https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#creating-an-eclipse-development-environment

Then as @mickaelistria  mentioned starting with something small is probably the best, we have the [good first issue](https://github.com/eclipse-platform/eclipse.platform.ui/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) label on all repositories.

Beside fixing actual bugs or implement features, looking for reproducers or adding tests is always welcome, if anything is unclear / missing don't hesitate to ask so we can improve documentation.

And yes there is some complexity, but this is not only for newcommers so don't be to concerned :-)

**Reply by mickaelistria** [reply]
> at the very minimum, you could setup and SDK like described here: 
https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#creating-an-eclipse-development-environment

I personally find this approach of setting up a whole IDE dedicated to entire Platform development not suitable for newcomers who do know about PDE and just want to fix one thing. Just telling them to use the SDK as target platform (as it's usually the case by default) or to deal with a target-platform which has recent Eclipse Platform; and to import in the workspace only the very one bundle they want to tweak is much less overwhelming and allows to get faster towards an actual change. It also allows them to work simultaneously on their own plugins/products and Platform without switching IDE/workspace.

**Reply by laeubi** [reply]
The problem is that platform often fast diverge from what is released and the you get strange and hard to understand errors and maybe end up in pulling in effectively everything. The next problem is that often things are not very well separated and usual Eclipse install missing the sources.

Anyways there are different options and ways people work, so the best would be to add a new section in the contribution guide describe the steps and possibly add some guidance (setting target platform, how to get the source, how to find out the right repository and so on).

**Reply by thahnen** [reply]
Every time I committed now to any project providing the Oomph magic stuff, I used it via the Eclipse installer while having a fork of the repo by the side. Then used patch files to transfer stuff from the actual "installation" to my fork, not the best approach I guess but I found it to be good enough without tinkering with this Oomph magic (I still have no idea how this sorcery works).

I like that the platform as well as PDE have their own dedicated installers and most of the time I also update them daily like my personal environment based on iBuilds ^^

**Reply by thahnen** [reply]
I wasn't aware of the `good-first-issue` labels until now; thanks for the heads up!

**Reply by mickaelistria** [reply]
All that is relatively basic GitHub and PDE usage which both already have great documentation; and actually developing to Platform is also basic PDE usage (this is not being said enough). I would rather enjoy removing a lot of extra (and IMO confusing and overwhelming) documentation to reference well-established one than adding more.

**Reply by laeubi** [reply]
I don't think it is that "basic" for newcommers, but if there is already good documentation adding a link in the contribution section would help for sure.

**Reply by mickaelistria** [reply]
> I don't think it is that "basic" for newcommers

It depends on how you define newcomers; IMO for a the Platform project you do want and need "newcomers" to be comfortable with Git and PDE. People who are not comfortable with those cannot be immediately helpful and will be a burden more than an opportunity for the project. The only good answer to them is to start by doing some tutorial about Git and plugin development and come back once they're ready.
I strongly believe that if we do state this pre-requisite skills (and link to resources allowing to gather those skills) and then design all our workflows for people who have such skills, then the "How to contribute to Eclipse project" part becomes almost empty, making contributors realize it's not a project any more complex than any other Eclipse plugin/product project.
The current focus on a "starting from scratch" experience hides the fact that Platform is as simple as any other Eclipse project and make it seem much more complex than it actually is to contribute.

**Reply by HannesWell** [reply]
> > at the very minimum, you could setup and SDK like described here:
> > https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#creating-an-eclipse-development-environment
> 
> I personally find this approach of setting up a whole IDE dedicated to entire Platform development not suitable for newcomers who do know about PDE and just want to fix one thing. Just telling them to use the SDK as target platform (as it's usually the case by default) or to deal with a target-platform which has recent Eclipse Platform; and to import in the workspace only the very one bundle they want to tweak is much less overwhelming and allows to get faster towards an actual change.

Initially this _can_ be fastest way, with the limitations Christoph mentioned. But it also has the drawback, that if one continues contributing one will one day hit problems that are too complicated, e.g. span multiple projects, to be reasonably solvable in such setup. And then one has to rework the local development setup. Therefore I think it's better to start with a 'scalable' solution immediately and make it simple when it's small. And I think the Oomph installer with corresponding setups can be such a solution.
And when one follows for example the guide at https://github.com/eclipse-ide/.github/blob/main/CONTRIBUTING.md, there are not many more step than if one imports a plugin project into an existing workspace using the running platform as target.

**Reply by vogella** [reply]
I have also give kudos to Sonarlint. Thanks @thahnen for working on it. I once coached a very senior Google developer to become a Eclipse committer and even he was super impressed how many issues were found in his code by Sonarlint.

**Reply by thahnen** [reply]
Hi @vogella, This is very Kind for you to say!
If I wouldn‘t find it very helpful I wouldn‘t be working on it and pouring Most of my Energy into keeping it alive and improve it for very loyal and thankful Eclipse users 🥳
And to a certain degree I get to do it for Eclipse as well ^^
